packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt

```diff
@@ -512,15 +510,146 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
 
 
     @VisibleForTesting
-    fun deserializeNestedModels(serializedModelData: Map<String, Any>): Map<String, Any> {
-        return serializedModelData.mapValues {
-            if (it.value is Map<*, *>) {
-                SerializedModel.builder()
-                    .serializedData(deserializeNestedModels(it.value as HashMap<String, Any>))
-                    .modelSchema(null)
+    fun deserializeNestedModel(serializedModelData: Map<String, Any?>, modelSchema: ModelSchema): Map<String, Any?> {
+        val result = mutableMapOf<String, Any?>()
+
+        // iterate over schema fields to deserialize each field value
+        for ((key, field) in modelSchema.fields.entries) {
+            // ignore if serializedModelData doesn't contain value for the current field key
+            if (!serializedModelData.containsKey(key)) {
+                continue
+            }
+            val fieldSerializedData = serializedModelData[key]
+
+            // if the field serialized value is null
+            // assign null to this field in the deserialized map
+            if (fieldSerializedData == null) {
+                result[key] = fieldSerializedData
+                continue
+            }
+
+            if (field.isModel) {
+                // ignore field if the field doesn't have valid schema in ModelProvider
+                val fieldModelSchema = modelProvider.modelSchemas()[field.targetType] ?: continue
+                result[key] = SerializedModel.builder()
+                    .serializedData(deserializeNestedModel(fieldSerializedData as Map<String, Any?>, fieldModelSchema))
+                    .modelSchema(fieldModelSchema)
                     .build()
-            } else
-                it.value
+            } else if (field.isCustomType) {
+                // ignore field if the field doesn't have valid schema in ModelProvider
+                val fieldCustomTypeSchema = modelProvider.customTypeSchemas()[field.targetType] ?: continue
+                val deserializedCustomType = getDeserializedCustomTypeField(
+                    fieldCustomTypeSchema = fieldCustomTypeSchema,
+                    isFieldArray = field.isArray,
+                    listOfSerializedData = if (field.isArray) fieldSerializedData as List<Map<String, Any?>> else null,
```

Comment:
What about nullable lists, i.e. `List<CustomType?>` ?

Reply:
I think nullable lists is still not supported by codegen atm, there is a nullable list work is going on in parallel but that depends on this PR to run in iOS. So I didn't include nullable element when casting to `List`, I could change the typing here, should be trivial.